chrysom/listenerClient.go

```diff
@@ -38,88 +39,108 @@ const (
-	PullInterval time.Duration
-
-	// Logger to be used by the client.
-	// (Optional). By default a no op logger will be used.
-	Logger *zap.Logger
+	// Config configures the ancla client and its listeners.
+	Config BasicClientConfig
+	// PollsTotalCounter measures the number of polls (and their success/failure outcomes) to fetch new items.
+	PollsTotalCounter *prometheus.CounterVec `name:"chrysom_polls_total"`
+	// Reader is the DB interface used to fetch new items using `GeItems`.
+	Reader Reader
+	// GetLogger returns a logger from the given context.
+	GetLogger GetLogger
+	// SetLogger embeds the `Listener.logger` in outgoing request contexts for `Listener.Update` calls.
+	SetLogger SetLogger
 }
 
 // ListenerClient is the client used to poll Argus for updates.
 type ListenerClient struct {
 	observer  *observerConfig
-	logger    *zap.Logger
+	getLogger func(context.Context) *zap.Logger
 	setLogger func(context.Context, *zap.Logger) context.Context
 	reader    Reader
 }
 
 type observerConfig struct {
-	listener     Listener
-	ticker       *time.Ticker
-	pullInterval time.Duration
-	measures     *Measures
-	shutdown     chan struct{}
-	state        int32
+	listener          Listener
+	ticker            *time.Ticker
+	pullInterval      time.Duration
+	pollsTotalCounter *prometheus.CounterVec
+
+	shutdown chan struct{}
+	state    int32
 }
 
-// NewListenerClient creates a new ListenerClient to be used to poll Argus
-// for updates.
-func NewListenerClient(config ListenerClientConfig,
-	setLogger func(context.Context, *zap.Logger) context.Context,
-	measures *Measures, r Reader,
-) (*ListenerClient, error) {
-	err := validateListenerConfig(&config)
+// ProvideListenerClient provides a new ListenerClient.
+func ProvideListenerClient(in ListenerClientIn) (*ListenerClient, error) {
+	client, err := NewListenerClient(in.Listener, in.GetLogger, in.SetLogger, in.Config.PullInterval, in.PollsTotalCounter, in.Reader)
 	if err != nil {
-		return nil, err
+		return nil, errors.Join(err, errFailedConfig)
 	}
-	if measures == nil {
-		return nil, ErrNilMeasures
+
+	return client, nil
+}
+
+func ProvideDefaultListenerReader(client *BasicClient) Reader {
+	return client
+}
+
+// ProvideListenerClient builds the Argus listener client service from the given configuration.
+// It allows adding watchers for the internal subscription state. Call the returned
+// function when you are done watching for updates.
+func NewListenerClient(listener Listener,
+	getLogger func(context.Context) *zap.Logger,
+	setLogger func(context.Context, *zap.Logger) context.Context,
+	pullInterval time.Duration, pollsTotalCounter *prometheus.CounterVec, reader Reader) (*ListenerClient, error) {
+	if listener == nil {
+		return nil, ErrNoListenerProvided
+	}
+	if pullInterval == 0 {
+		pullInterval = defaultPullInterval
 	}
 	if setLogger == nil {
 		setLogger = func(ctx context.Context, _ *zap.Logger) context.Context {
 			return ctx
 		}
 	}
-	if r == nil {
+	if reader == nil {
```

Comment:
code moved from `validateListenerConfig` into `NewListenerClient` in order to simplify the structure and flow of startup